mobile_verifier/src/promotion_reward.rs

```diff
@@ -0,0 +1,565 @@
+        carrier_client: &CarrierServiceClient,
+    ) -> anyhow::Result<Self> {
+        // mobile_config::client::ClientError::UnknownServiceProvider(_)
+        let (service_provider, validity) = match carrier_client
+            .payer_key_to_service_provider(
+                &promotion_reward.carrier_pub_key.to_string(), // ????
+            )
+            .await
+        {
+            Ok(service_provider) => (Some(service_provider), PromotionRewardStatus::Valid),
+            Err(mobile_config::client::ClientError::UnknownServiceProvider(_)) => {
+                (None, PromotionRewardStatus::InvalidCarrierKey)
+            }
+            Err(x) => return Err(x.into()),
+        };
+        // TODO: Check if entity is rewardable
+        Ok(Self {
+            validity,
+            promotion_reward,
+            service_provider,
+        })
+    }
+
+    fn validate_promotion_rewards<'a>(
+        promotion_rewards: impl Stream<Item = PromotionReward> + 'a,
+        carrier_client: &'a CarrierServiceClient,
+    ) -> impl Stream<Item = anyhow::Result<Self>> + 'a {
+        promotion_rewards.then(move |promotion_reward| async move {
+            Self::validate(promotion_reward, carrier_client).await
+        })
+    }
+
+    fn is_valid(&self) -> bool {
+        matches!(self.validity, PromotionRewardStatus::Valid)
+    }
+
+    async fn write(
+        &self,
+        promotion_rewards: &FileSinkClient<VerifiedPromotionRewardV1>,
+    ) -> anyhow::Result<()> {
+        promotion_rewards
+            .write(
+                VerifiedPromotionRewardV1 {
+                    report: Some(PromotionRewardIngestReportV1 {
+                        received_timestamp: self
+                            .promotion_reward
+                            .received_timestamp
+                            .encode_timestamp_millis(),
+                        report: Some(self.promotion_reward.clone().into()),
+                    }),
+                    status: self.validity as i32,
+                    timestamp: Utc::now().encode_timestamp_millis(),
+                },
+                &[("validity", self.validity.as_str_name())],
+            )
+            .await?;
+        Ok(())
+    }
+
+    async fn save(self, transaction: &mut Transaction<'_, Postgres>) -> anyhow::Result<()> {
+        let (entity, entity_type) = match self.promotion_reward.entity {
+            Entity::SubscriberId(sub_id) => (sub_id.to_string(), EntityType::Subscriber),
+            Entity::GatewayKey(gk) => (gk.to_string(), EntityType::Gateway),
+        };
+        sqlx::query(
+            r#"
+            INSERT INTO promotion_rewards (time_of_reward, entity, entity_type, service_provider, shares)
+            VALUES ($1, $2, $3, $4, $5)
+            ON CONFLICT DO NOTHING
+            "#
+        )
+            .bind(self.promotion_reward.timestamp)
+            .bind(entity)
+            .bind(entity_type)
+            .bind(self.service_provider.map(|x| x as i64))
```

Comment:
`service_provider` is a `NOT NULL` column in the migration. 
I'm not sure if it makes sense to check earlier that the service provider exists, or to remove the option from this struct.